### PR TITLE
3417 add install agent plugin archive

### PR DESCRIPTION
**Pull request by ilija-lazoroski**
# What does this PR do?

Fixes part of #3417 .

Adds `IAgentPluginService.install_agent_plugin_archice`

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [ ] Was the CHANGELOG.md updated to reflect the changes?
* [ ] Was the documentation framework updated to reflect the changes?
* [ ] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Do all unit tests pass?
* [x] Do all end-to-end tests pass?
* [x] Any other testing performed?
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [ ] If applicable, add screenshots or log transcripts of the feature working

### CODE REVIEW

File: monkey/monkey_island/cc/services/agent_plugin_service/agent_plugin_service.py [hunk 1]
```diff
@@ -30,3 +34,14 @@ def get_all_plugin_configuration_schemas(
 
     def get_all_plugin_manifests(self) -> Dict[AgentPluginType, Dict[str, AgentPluginManifest]]:
         return self._agent_plugin_repository.get_all_plugin_manifests()
+
+    def install_agent_plugin_archive(self, agent_plugin_archive: bytes):
+        with self._lock:
+            os_agent_plugins = parse_plugin(io.BytesIO(agent_plugin_archive))
+            for operating_system, agent_plugin in os_agent_plugins.items():
+                self._agent_plugin_repository.remove_agent_plugin(
+                    None if len(os_agent_plugins) == 1 else operating_system,
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/services/agent_plugin_service/agent_plugin_service.py:43]: This logic is needlessly complicated and could also lead to an inconsistent system state.

```suggestion
                    None
```

File: monkey/monkey_island/cc/services/agent_plugin_service/i_agent_plugin_service.py [hunk 2]
```diff
@@ -44,3 +44,14 @@ def get_all_plugin_manifests(self) -> Dict[AgentPluginType, Dict[str, AgentPlugi
...
         """
         pass
+
+    @abstractmethod
+    def install_agent_plugin_archive(self, agent_plugin_archive: bytes):
+        """
+        Install plugin archive
+
+        :param agent_plugin_archive: The archive of the plugin
+        :raises RemovalError: If an error occus while attempting to remove the plugin
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/services/agent_plugin_service/i_agent_plugin_service.py:54]:
```suggestion
        :raises RemovalError: If an error occus while attempting to uninstall a previous version of the plugin
```

File: monkey/monkey_island/cc/services/agent_plugin_service/i_agent_plugin_service.py [hunk 3]
```diff
@@ -44,3 +44,14 @@ def get_all_plugin_manifests(self) -> Dict[AgentPluginType, Dict[str, AgentPlugi
...
+        :raises StorageError: If an error occurs while attempting to store the plugin
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/services/agent_plugin_service/i_agent_plugin_service.py:55]:
```suggestion
        :raises StorageError: If an error occurs while attempting to install the plugin
```

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/tests/unit_tests/monkey_island/cc/services/agent_plugin_service/test_agent_plugin_service.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+)
+
+
+@pytest.fixture
+def agent_plugin_repository() -> IAgentPluginRepository:
+    return MagicMock(spec=IAgentPluginRepository)
+
+
+@pytest.fixture
+def agent_plugin_servce(agent_plugin_repository) -> IAgentPluginService:
```
Comment (VakarisZ) [inline monkey/tests/unit_tests/monkey_island/cc/services/agent_plugin_service/test_agent_plugin_service.py:23]: typo

File: monkey/monkey_island/cc/services/agent_plugin_service/agent_plugin_service.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mssalvatore) [reply monkey/monkey_island/cc/services/agent_plugin_service/agent_plugin_service.py:43]: Yeah, I don't love it. The semantics are defined such that `None` is all operating systems. Simply invoking this with keyword args and leaving operating system out would make the call much more intuitive.

`remove_agent_plugin(Type=agent_plugin.plugin_manifest.plugin_type, name=agent_plugin.plugin_manifest.name)`

For consistency, order should be type, name.

File: monkey/tests/unit_tests/monkey_island/cc/services/agent_plugin_service/test_agent_plugin_service.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+)
+def test_agent_plugin_service__install_agent_plugin_archive(
+    plugin_data_dir: Path,
+    plugin_os: OperatingSystem,
+    plugin_path: str,
+    agent_plugin_repository: IAgentPluginRepository,
+    agent_plugin_servce: IAgentPluginService,
+    build_agent_plugin_tar_with_source_tar: Callable[[Path], BinaryIO],
+):
+    agent_plugin_tar = build_agent_plugin_tar_with_source_tar(plugin_data_dir / plugin_path)
```
[the remark this answers is not in the document]
Comment (mssalvatore) [reply monkey/tests/unit_tests/monkey_island/cc/services/agent_plugin_service/test_agent_plugin_service.py:42]: I think we do for some tests. If we do that, we'll have to rebuild them if the interface changes. This method lets the tests ask specifically for what they want.

I don't mean to say "no", I'm just laying out pros and cons.

**Pull request merged**